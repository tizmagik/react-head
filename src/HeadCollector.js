import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Provider } from './headTagsContext';

export default class HeadCollector extends Component {
  static propTypes = {
    headTags: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
    children: PropTypes.node.isRequired,
  };

  headTags = {
    add: tag => this.props.headTags.push(tag),
  };

  render() {
    return <Provider value={this.headTags}>{React.Children.only(this.props.children)}</Provider>;
  }
}
