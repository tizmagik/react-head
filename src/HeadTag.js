import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import buildSelector from './buildSelector';

export default class HeadTag extends Component {
  static contextTypes = {
    reactHeadTags: PropTypes.object,
  };

  static propTypes = {
    tag: PropTypes.string,
  };

  static defaultProps = {
    tag: 'meta',
  };

  state = {
    canUseDOM: false,
  };

  componentDidMount() {
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({ canUseDOM: true });

    const { tag, children, ...rest } = this.props; // eslint-disable-line react/prop-types
    const ssrTags = document.head.querySelector(`${tag}${buildSelector(rest)}[data-rh=""]`);

    /* istanbul ignore else */
    if (ssrTags) {
      ssrTags.remove();
    }
  }

  render() {
    const { tag: Tag, ...rest } = this.props;

    if (this.state.canUseDOM) {
      const Comp = <Tag {...rest} />;
      return ReactDOM.createPortal(Comp, document.head);
    }

    // on client we don't require HeadCollector
    if (this.context.reactHeadTags) {
      const ServerComp = <Tag data-rh="" {...rest} />;
      this.context.reactHeadTags.add(ServerComp);
    }

    return null;
  }
}
