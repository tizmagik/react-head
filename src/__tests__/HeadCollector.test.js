import React from 'react';
import { render } from 'enzyme';
import HeadCollector from '../HeadCollector';
import HeadTag from '../HeadTag';

describe('HeadCollector', () => {
  it('adds HeadTags to given array from component tree', () => {
    const arr = [];
    render(
      <HeadCollector headTags={arr}>
        <div>
          <HeadTag tag="tag1" name="name1" another="value1" />
          <HeadTag tag="tag2" test="test2" third="value2" />
        </div>
      </HeadCollector>
    );

    expect(arr.length).toBe(2);
    expect(arr).toMatchSnapshot();
  });
});
