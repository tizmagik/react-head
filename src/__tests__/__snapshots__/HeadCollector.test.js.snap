// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadCollector adds HeadTags to given array from component tree 1`] = `
Array [
  <tag1
    another="value1"
    data-rh=""
    name="name1"
  />,
  <tag2
    data-rh=""
    test="test2"
    third="value2"
  />,
]
`;
